File: scripts/identify-super-enhancers.ts
/**
 * Super-Enhancer Identification from BigWig files.
 *
 * Implements the ROSE algorithm concept:
 * 1. Identify peaks (regions with signal above threshold)
 * 2. Stitch nearby peaks within 12.5kb
 * 3. Rank by total signal
 * 4. Find inflection point (hockey stick) to separate SE from typical enhancers
 *
 * Input: H3K27ac BigWig (primary) + MED1 BigWig (optional validation)
 * Output: BED file with Super-Enhancer coordinates
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ============================================================================
// Configuration
// ============================================================================

const DEFAULT_H3K27AC = 'D:/ПАРСИНГ НАУЧНЫХ НОВОСТЕЙ/data/inputs/histone/H3K27ac_GM12878.bw';
const DEFAULT_MED1 = 'D:/ПАРСИНГ НАУЧНЫХ НОВОСТЕЙ/data/inputs/med1/MED1_GM12878_Rep1.bw';
const OUTPUT_DIR = path.join(__dirname, '..', 'results');

const RESOLUTION = 1000;  // 1kb bins
const STITCH_DISTANCE = 12500;  // 12.5kb in bp
const PEAK_THRESHOLD_PERCENTILE = 75;  // Lower threshold
const MIN_ENHANCER_SIZE = 2000;  // 2kb minimum

// Main chromosomes to process
const MAIN_CHROMS = [
    'chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8',
    'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',
    'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX'
];

// ============================================================================
// Types
// ============================================================================

interface Peak {
    chrom: string;
    start: number;
    end: number;
    signal: number;
}

interface Enhancer extends Peak {
    type: 'SE' | 'TE';
    rank?: number;
}

// ============================================================================
// BigWig Processing
// ============================================================================

async function processChromosome(bwPath: string, chrom: string, chromSize: number): Promise<Peak[]> {
    const { BigWig } = await import('@gmod/bbi');
    const bw = new BigWig({ path: bwPath });
    await bw.getHeader();

    const peaks: Peak[] = [];
    const GAP_TOLERANCE = 500;  // Allow gaps up to 500bp between high-signal features

    // Process in chunks to handle large chromosomes
    const chunkSize = 10_000_000;  // 10Mb chunks

    for (let chunkStart = 0; chunkStart < chromSize; chunkStart += chunkSize) {
        const chunkEnd = Math.min(chunkStart + chunkSize, chromSize);

        try {
            const features = await bw.getFeatures(chrom, chunkStart, chunkEnd, { scale: 1 });

            if (features.length === 0) continue;

            // Calculate threshold for this chunk (use 90th percentile for stricter peaks)
            const scores = features.map((f: any) => f.score || 0).filter((s: number) => s > 0);
            if (scores.length === 0) continue;

            scores.sort((a: number, b: number) => a - b);
            const thresholdIdx = Math.floor(scores.length * PEAK_THRESHOLD_PERCENTILE / 100);
            const threshold = scores[thresholdIdx] || 0;

            // Collect high-signal features
            const highSignalFeatures: Array<{ start: number; end: number; score: number }> = [];
            for (const feat of features as any[]) {
                if (feat.score > threshold) {
                    highSignalFeatures.push({
                        start: feat.start,
                        end: feat.end,
                        score: feat.score
                    });
                }
            }

            if (highSignalFeatures.length === 0) continue;

            // Sort by start position
            highSignalFeatures.sort((a, b) => a.start - b.start);

            // Merge nearby high-signal features into peaks (allow gaps up to GAP_TOLERANCE)
            let peakStart = highSignalFeatures[0].start;
            let peakEnd = highSignalFeatures[0].end;
            let peakSignal = highSignalFeatures[0].score * (highSignalFeatures[0].end - highSignalFeatures[0].start);

            for (let i = 1; i < highSignalFeatures.length; i++) {
                const feat = highSignalFeatures[i];

                if (feat.start - peakEnd <= GAP_TOLERANCE) {
                    // Extend peak
                    peakEnd = feat.end;
                    peakSignal += feat.score * (feat.end - feat.start);
                } else {
                    // Save current peak if large enough
                    if (peakEnd - peakStart >= MIN_ENHANCER_SIZE) {
                        peaks.push({
                            chrom,
                            start: peakStart,
                            end: peakEnd,
                            signal: peakSignal
                        });
                    }
                    // Start new peak
                    peakStart = feat.start;
                    peakEnd = feat.end;
                    peakSignal = feat.score * (feat.end - feat.start);
                }
            }

            // Don't forget last peak
            if (peakEnd - peakStart >= MIN_ENHANCER_SIZE) {
                peaks.push({
                    chrom,
                    start: peakStart,
                    end: peakEnd,
                    signal: peakSignal
                });
            }
        } catch (e) {
            // Skip this chunk on error
        }
    }

    return peaks;
}

function stitchPeaks(peaks: Peak[]): Peak[] {
    if (peaks.length === 0) return [];

    // Sort by chromosome and start position
    const sorted = [...peaks].sort((a, b) => {
        if (a.chrom !== b.chrom) return a.chrom.localeCompare(b.chrom);
        return a.start - b.start;
    });

    const stitched: Peak[] = [];
    let current = { ...sorted[0] };

    for (let i = 1; i < sorted.length; i++) {
        const peak = sorted[i];

        if (peak.chrom === current.chrom && peak.start - current.end <= STITCH_DISTANCE) {
            // Stitch together
            current.end = peak.end;
            current.signal += peak.signal;
        } else {
            // Save current and start new
            stitched.push(current);
            current = { ...peak };
        }
    }

    // Don't forget the last one
    stitched.push(current);

    return stitched;
}

function findInflectionPoint(signals: number[]): number {
    if (signals.length < 3) return signals[0] || 0;

    // Sort descending
    const sorted = [...signals].sort((a, b) => b - a);
    const maxSignal = sorted[0];
    if (maxSignal === 0) return 0;

    // Hockey stick method: find point with max distance from diagonal
    let maxDistance = 0;
    let inflectionIdx = 0;

    for (let i = 0; i < sorted.length; i++) {
        const x = i / sorted.length;
        const y = sorted[i] / maxSignal;
        const diagonal = 1 - x;
        const distance = Math.abs(y - diagonal);

        if (distance > maxDistance) {
            maxDistance = distance;
            inflectionIdx = i;
        }
    }

    return sorted[inflectionIdx];
}

// ============================================================================
// Main
// ============================================================================

async function main() {
    console.log('');
    console.log('█'.repeat(70));
    console.log('  SUPER-ENHANCER IDENTIFICATION (ROSE-like algorithm)');
    console.log('█'.repeat(70));
    console.log('');

    const h3k27acPath = DEFAULT_H3K27AC;
    const med1Path = DEFAULT_MED1;
    const outputBed = path.join(OUTPUT_DIR, 'super_enhancers_GM12878.bed');

    console.log(`H3K27ac: ${h3k27acPath}`);
    console.log(`MED1:    ${med1Path}`);
    console.log(`Output:  ${outputBed}`);
    console.log('');

    // Check files exist
    if (!fs.existsSync(h3k27acPath)) {
        console.error(`ERROR: H3K27ac file not found: ${h3k27acPath}`);
        process.exit(1);
    }

    // Get chromosome sizes
    const { BigWig } = await import('@gmod/bbi');
    const bw = new BigWig({ path: h3k27acPath });
    const header = await bw.getHeader();
    const chromSizes = new Map<string, number>();

    // refsByNumber contains {name, id, length} for each chromosome
    for (const [, info] of Object.entries(header.refsByNumber as Record<string, { name: string; id: number; length: number }>)) {
        chromSizes.set(info.name, info.length);
    }

    console.log(`Found ${chromSizes.size} chromosomes`);
    console.log('');

    // Process each chromosome
    const allPeaks: Peak[] = [];

    for (const chrom of MAIN_CHROMS) {
        const size = chromSizes.get(chrom);
        if (!size) {
            console.log(`${chrom}: not found in BigWig`);
            continue;
        }

        process.stdout.write(`${chrom}...`);
        const peaks = await processChromosome(h3k27acPath, chrom, size);
        allPeaks.push(...peaks);
        console.log(` ${peaks.length} peaks`);
    }

    console.log('');
    console.log(`Total peaks before stitching: ${allPeaks.length}`);

    // Stitch peaks
    const stitched = stitchPeaks(allPeaks);
    console.log(`After stitching (${STITCH_DISTANCE / 1000}kb): ${stitched.length} regions`);

    if (stitched.length === 0) {
        console.log('WARNING: No enhancers found!');
        process.exit(1);
    }

    // Find inflection point
    const signals = stitched.map(p => p.signal);
    const inflectionSignal = findInflectionPoint(signals);
    console.log(`Inflection point signal: ${inflectionSignal.toFixed(2)}`);

    // Classify
    const enhancers: Enhancer[] = stitched.map(p => ({
        ...p,
        type: p.signal >= inflectionSignal ? 'SE' as const : 'TE' as const
    }));

    // Sort by signal
    enhancers.sort((a, b) => b.signal - a.signal);

    // Add ranks
    enhancers.forEach((e, i) => e.rank = i + 1);

    const superEnhancers = enhancers.filter(e => e.type === 'SE');
    const typicalEnhancers = enhancers.filter(e => e.type === 'TE');

    console.log('');
    console.log(`Super-Enhancers (SE): ${superEnhancers.length}`);
    console.log(`Typical Enhancers (TE): ${typicalEnhancers.length}`);

    // Write BED file for Super-Enhancers
    const bedLines: string[] = [
        '#chrom\tstart\tend\tname\tscore\tstrand\tsignal\trank'
    ];

    for (const se of superEnhancers) {
        const name = `SE_${se.rank}_${se.chrom}`;
        const score = Math.min(1000, Math.round(se.signal / 100));
        bedLines.push(`${se.chrom}\t${se.start}\t${se.end}\t${name}\t${score}\t.\t${se.signal.toFixed(2)}\t${se.rank}`);
    }

    fs.writeFileSync(outputBed, bedLines.join('\n'), 'utf-8');
    console.log(`\nSuper-Enhancers written to: ${outputBed}`);

    // Write all enhancers
    const allBed = outputBed.replace('.bed', '_all.bed');
    const allLines: string[] = [
        '#chrom\tstart\tend\tname\tscore\tstrand\tsignal\ttype\trank'
    ];

    for (const e of enhancers) {
        const name = `${e.type}_${e.rank}_${e.chrom}`;
        const score = Math.min(1000, Math.round(e.signal / 100));
        allLines.push(`${e.chrom}\t${e.start}\t${e.end}\t${name}\t${score}\t.\t${e.signal.toFixed(2)}\t${e.type}\t${e.rank}`);
    }

    fs.writeFileSync(allBed, allLines.join('\n'), 'utf-8');
    console.log(`All enhancers written to: ${allBed}`);

    // Print top 20
    console.log('');
    console.log('═'.repeat(75));
    console.log('TOP 20 SUPER-ENHANCERS');
    console.log('═'.repeat(75));
    console.log(`${'Rank'.padEnd(6)}${'Chrom'.padEnd(8)}${'Start'.padEnd(12)}${'End'.padEnd(12)}${'Length'.padEnd(10)}${'Signal'.padEnd(14)}`);
    console.log('-'.repeat(75));

    for (const se of superEnhancers.slice(0, 20)) {
        const length = se.end - se.start;
        console.log(
            `${(se.rank || 0).toString().padEnd(6)}` +
            `${se.chrom.padEnd(8)}` +
            `${se.start.toLocaleString().padEnd(12)}` +
            `${se.end.toLocaleString().padEnd(12)}` +
            `${(length / 1000).toFixed(1).padStart(6)}kb   ` +
            `${se.signal.toFixed(1).padEnd(14)}`
        );
    }

    // Validate top SE with MED1
    if (fs.existsSync(med1Path)) {
        console.log('');
        console.log('═'.repeat(75));
        console.log('MED1 SIGNAL VALIDATION (Top 10 SE)');
        console.log('═'.repeat(75));

        const med1Bw = new BigWig({ path: med1Path });
        await med1Bw.getHeader();

        for (const se of superEnhancers.slice(0, 10)) {
            try {
                const feats = await med1Bw.getFeatures(se.chrom, se.start, se.end, { scale: 1 });
                let sum = 0;
                for (const f of feats as any[]) {
                    sum += (f.score || 0) * (f.end - f.start);
                }
                console.log(`SE_${se.rank} (${se.chrom}:${se.start.toLocaleString()}-${se.end.toLocaleString()}): MED1 = ${sum.toFixed(2)}`);
            } catch {
                console.log(`SE_${se.rank}: MED1 unavailable`);
            }
        }
    }

    console.log('');
    console.log('═'.repeat(75));
    console.log(`COMPLETE: ${superEnhancers.length} Super-Enhancers identified`);
    console.log('═'.repeat(75));
}

main().catch(err => {
    console.error('Fatal error:', err);
    process.exit(1);
});

File: scripts/identify-super-enhancers.py
#!/usr/bin/env python3
"""
Super-Enhancer Identification from BigWig files.

This script implements the ROSE algorithm concept:
1. Identify peaks (regions with signal above threshold)
2. Stitch nearby peaks within 12.5kb
3. Rank by total signal
4. Find inflection point (hockey stick) to separate SE from typical enhancers

Input: H3K27ac BigWig (primary) + MED1 BigWig (optional validation)
Output: BED file with Super-Enhancer coordinates
"""

import sys
import os
import argparse
import numpy as np

def check_dependencies():
    """Check if pyBigWig is installed."""
    try:
        import pyBigWig
        return True
    except ImportError:
        print("ERROR: pyBigWig not installed. Install with: pip install pyBigWig")
        return False

def load_bigwig_signal(bw_path, chrom, start, end, resolution=1000):
    """Load signal from BigWig file."""
    import pyBigWig
    bw = pyBigWig.open(bw_path)

    # Get values
    values = bw.values(chrom, start, end)
    bw.close()

    if values is None:
        return np.zeros((end - start) // resolution)

    # Convert to numpy and handle NaN
    values = np.array(values)
    values = np.nan_to_num(values, nan=0.0)

    # Bin the values
    n_bins = (end - start) // resolution
    binned = np.zeros(n_bins)
    for i in range(n_bins):
        bin_start = i * resolution
        bin_end = min((i + 1) * resolution, len(values))
        if bin_start < len(values):
            binned[i] = np.mean(values[bin_start:bin_end])

    return binned

def call_peaks(signal, threshold_percentile=90, min_width=2):
    """Call peaks based on signal threshold."""
    threshold = np.percentile(signal[signal > 0], threshold_percentile) if np.any(signal > 0) else 0

    peaks = []
    in_peak = False
    peak_start = 0

    for i, val in enumerate(signal):
        if val > threshold and not in_peak:
            in_peak = True
            peak_start = i
        elif val <= threshold and in_peak:
            in_peak = False
            if i - peak_start >= min_width:
                peaks.append((peak_start, i, np.sum(signal[peak_start:i])))

    # Handle peak at end
    if in_peak and len(signal) - peak_start >= min_width:
        peaks.append((peak_start, len(signal), np.sum(signal[peak_start:])))

    return peaks

def stitch_peaks(peaks, stitch_distance=12):
    """Stitch peaks within stitch_distance bins."""
    if not peaks:
        return []

    # Sort by start position
    peaks = sorted(peaks, key=lambda x: x[0])

    stitched = []
    current_start, current_end, current_signal = peaks[0]

    for start, end, signal in peaks[1:]:
        if start - current_end <= stitch_distance:
            # Stitch
            current_end = end
            current_signal += signal
        else:
            # Save current and start new
            stitched.append((current_start, current_end, current_signal))
            current_start, current_end, current_signal = start, end, signal

    # Don't forget the last one
    stitched.append((current_start, current_end, current_signal))

    return stitched

def find_inflection_point(signals):
    """Find the inflection point using the hockey stick method."""
    if len(signals) < 3:
        return 0

    # Sort signals
    sorted_signals = np.sort(signals)[::-1]

    # Normalize to 0-1 range
    x = np.arange(len(sorted_signals)) / len(sorted_signals)
    y = sorted_signals / sorted_signals[0] if sorted_signals[0] > 0 else sorted_signals

    # Find point with maximum distance from diagonal
    diagonal = 1 - x
    distances = np.abs(y - diagonal)

    inflection_idx = np.argmax(distances)

    # Return the signal value at inflection point
    return sorted_signals[inflection_idx]

def identify_super_enhancers(bw_path, output_bed, genome='hg38', resolution=1000, stitch_kb=12.5):
    """Main function to identify super-enhancers."""
    import pyBigWig

    print(f"Loading BigWig: {bw_path}")
    bw = pyBigWig.open(bw_path)

    # Get chromosome sizes
    chroms = bw.chroms()
    bw.close()

    # Filter to main chromosomes
    main_chroms = [c for c in chroms.keys() if c.startswith('chr') and c[3:].isdigit() or c in ['chrX', 'chrY']]
    main_chroms = sorted(main_chroms, key=lambda x: (0, int(x[3:])) if x[3:].isdigit() else (1, x))

    print(f"Processing {len(main_chroms)} chromosomes...")

    all_enhancers = []
    stitch_bins = int(stitch_kb * 1000 / resolution)

    for chrom in main_chroms:
        chrom_size = chroms[chrom]

        # Load signal
        signal = load_bigwig_signal(bw_path, chrom, 0, chrom_size, resolution)

        if np.sum(signal) == 0:
            continue

        # Call peaks
        peaks = call_peaks(signal, threshold_percentile=85, min_width=2)

        # Stitch peaks
        stitched = stitch_peaks(peaks, stitch_distance=stitch_bins)

        # Convert to genomic coordinates
        for start_bin, end_bin, total_signal in stitched:
            start_bp = start_bin * resolution
            end_bp = end_bin * resolution
            all_enhancers.append({
                'chrom': chrom,
                'start': start_bp,
                'end': end_bp,
                'signal': total_signal,
                'length': end_bp - start_bp
            })

        print(f"  {chrom}: {len(stitched)} stitched regions")

    print(f"\nTotal enhancer regions: {len(all_enhancers)}")

    if not all_enhancers:
        print("No enhancers found!")
        return []

    # Find inflection point
    signals = np.array([e['signal'] for e in all_enhancers])
    inflection_signal = find_inflection_point(signals)

    print(f"Inflection point signal: {inflection_signal:.2f}")

    # Classify as Super-Enhancer or Typical Enhancer
    super_enhancers = []
    typical_enhancers = []

    for e in all_enhancers:
        if e['signal'] >= inflection_signal:
            e['type'] = 'SE'
            super_enhancers.append(e)
        else:
            e['type'] = 'TE'
            typical_enhancers.append(e)

    # Sort by signal (descending)
    super_enhancers = sorted(super_enhancers, key=lambda x: -x['signal'])

    print(f"\nSuper-Enhancers: {len(super_enhancers)}")
    print(f"Typical Enhancers: {len(typical_enhancers)}")

    # Write BED file
    with open(output_bed, 'w') as f:
        f.write("#chrom\tstart\tend\tname\tscore\tstrand\tsignal\ttype\n")

        for i, se in enumerate(super_enhancers):
            name = f"SE_{i+1}_{se['chrom']}"
            score = min(1000, int(se['signal']))
            f.write(f"{se['chrom']}\t{se['start']}\t{se['end']}\t{name}\t{score}\t.\t{se['signal']:.2f}\tSE\n")

    print(f"\nSuper-Enhancers written to: {output_bed}")

    # Also write all enhancers for reference
    all_bed = output_bed.replace('.bed', '_all.bed')
    with open(all_bed, 'w') as f:
        f.write("#chrom\tstart\tend\tname\tscore\tstrand\tsignal\ttype\n")

        all_sorted = sorted(all_enhancers, key=lambda x: -x['signal'])
        for i, e in enumerate(all_sorted):
            name = f"{e['type']}_{i+1}_{e['chrom']}"
            score = min(1000, int(e['signal']))
            f.write(f"{e['chrom']}\t{e['start']}\t{e['end']}\t{name}\t{score}\t.\t{e['signal']:.2f}\t{e['type']}\n")

    print(f"All enhancers written to: {all_bed}")

    # Print top 20 Super-Enhancers
    print("\n" + "="*70)
    print("TOP 20 SUPER-ENHANCERS")
    print("="*70)
    print(f"{'Rank':<6}{'Chrom':<8}{'Start':<12}{'End':<12}{'Length':<10}{'Signal':<12}")
    print("-"*70)

    for i, se in enumerate(super_enhancers[:20]):
        print(f"{i+1:<6}{se['chrom']:<8}{se['start']:<12}{se['end']:<12}{se['length']:<10}{se['signal']:<12.1f}")

    return super_enhancers

def main():
    parser = argparse.ArgumentParser(description='Identify Super-Enhancers from BigWig')
    parser.add_argument('--h3k27ac', required=True, help='H3K27ac BigWig file')
    parser.add_argument('--med1', help='MED1 BigWig file (optional, for validation)')
    parser.add_argument('--output', '-o', required=True, help='Output BED file')
    parser.add_argument('--resolution', type=int, default=1000, help='Bin resolution in bp (default: 1000)')
    parser.add_argument('--stitch', type=float, default=12.5, help='Stitch distance in kb (default: 12.5)')

    args = parser.parse_args()

    if not check_dependencies():
        sys.exit(1)

    if not os.path.exists(args.h3k27ac):
        print(f"ERROR: H3K27ac file not found: {args.h3k27ac}")
        sys.exit(1)

    super_enhancers = identify_super_enhancers(
        args.h3k27ac,
        args.output,
        resolution=args.resolution,
        stitch_kb=args.stitch
    )

    # If MED1 provided, validate overlap
    if args.med1 and os.path.exists(args.med1):
        print("\n" + "="*70)
        print("VALIDATING WITH MED1 SIGNAL")
        print("="*70)

        import pyBigWig
        bw = pyBigWig.open(args.med1)

        for i, se in enumerate(super_enhancers[:10]):
            try:
                med1_signal = bw.stats(se['chrom'], se['start'], se['end'], type='mean')[0]
                med1_signal = med1_signal if med1_signal else 0
                print(f"SE_{i+1} ({se['chrom']}:{se['start']}-{se['end']}): MED1 = {med1_signal:.2f}")
            except:
                pass

        bw.close()

if __name__ == '__main__':
    main()
